torchopt/transform/scale_by_rss.py

```diff
@@ -0,0 +1,156 @@
+
+from __future__ import annotations
+
+from typing import NamedTuple
+
+import torch
+
+from torchopt import pytree
+from torchopt.base import GradientTransformation
+from torchopt.transform.utils import tree_map_flat
+from torchopt.typing import OptState, Params, Updates
+
+
+__all__ = ['scale_by_rss']
+
+
+class ScaleByRssState(NamedTuple):
+    """State holding the sum of gradient squares to date."""
+
+    sum_of_squares: Updates
+
+
+def scale_by_rss(
+    initial_accumulator_value: float = 0.0,
+    eps: float = 1e-10,
+) -> GradientTransformation:
+    """Rescale updates by the root of the sum of all squared gradients to date.
+
+    References:
+        [Duchi et al, 2011](https://jmlr.org/papers/volume12/duchi11a/duchi11a.pdf)
+        [McMahan et al., 2010](https://arxiv.org/abs/1002.4908)
+
+    Args:
+        initial_accumulator_value: Starting value for accumulators, must be >= 0.
+        eps: A small floating point value to avoid zero denominator.
+
+    Returns:
+        An (init_fn, update_fn) tuple.
+    """
+    return _scale_by_rss(
+        initial_accumulator_value=initial_accumulator_value,
+        eps=eps,
+        already_flattened=False,
+    )
+
+
+def _scale_by_rss_flat(
+    initial_accumulator_value: float = 0.0,
+    eps: float = 1e-10,
+) -> GradientTransformation:
+    return _scale_by_rss(
+        initial_accumulator_value=initial_accumulator_value,
+        eps=eps,
+        already_flattened=True,
+    )
+
+
+def _scale_by_rss(
+    initial_accumulator_value: float = 0.0,
+    eps: float = 1e-10,
+    *,
+    already_flattened: bool = False,
+) -> GradientTransformation:
+    if already_flattened:  # noqa: SIM108
+        tree_map = tree_map_flat
+    else:
+        tree_map = pytree.tree_map  # type: ignore[assignment]
+
+    def init_fn(params: Params) -> OptState:
+        sum_of_squares = tree_map(
+            lambda t: torch.full_like(
+                t,
+                initial_accumulator_value,
+                memory_format=torch.preserve_format,
+            ),
+            params,
+        )
+        return ScaleByRssState(sum_of_squares=sum_of_squares)
+
+    def update_fn(
+        updates: Updates,
+        state: OptState,
+        params: Params | None = None,
+        inplace: bool = True,
+    ) -> tuple[Updates, OptState]:  # pylint: disable=unused-argument
+        del params
+        # sum_of_squares = tree_map(
+        #     lambda g, t: t + (g.conj() * g).real , updates, state.sum_of_squares
+        # )
+        # sum_of_squares = torch.addcmul(state.sum_of_squares, updates, updates, value=1)
+
+        sum_of_squares = tree_map(
+            lambda g, t: t.addcmul_(g, g, value=1.0),
+            updates,
+            state.sum_of_squares,
+        )
+
+        if inplace:
+
+            def f(t: torch.Tensor) -> torch.Tensor:
+                return torch.where(
+                    t > 0.0,
+                    torch.ones_like(t).div_(t.sqrt_().add(eps)),
```

Comment:
t will be modified by calling t.sqrt_; change: t.sqrt_().add(eps) -> t.sqrt().add_(eps)

Reply:
check others places, plz